Intern/rayx-ui/src/Triangulation/Triangulate.cpp

```diff
@@ -2,128 +2,423 @@
 
 #include <glm/gtc/matrix_transform.hpp>
 #include <glm/gtx/string_cast.hpp>
+#include <map>
 #include <memory>
+#include <queue>
+#include <set>
 
 #include "Colors.h"
 #include "Debug/Debug.h"
 #include "GeometryUtils.h"
 #include "Shader/Constants.h"
 #include "Triangulation/TraceTriangulation.h"
 
+struct Point2D {
+    double x;
+    double y;
+
+    int operator<(Point2D other) const { return y < other.y || (y == other.y && x > other.x); }
+    int operator>(Point2D other) const { return y > other.y || (y == other.y && x < other.x); }
+
+    int operator==(Point2D other) const { return x == other.x && y == other.y; }
+};
+
+// A simple polygon is a list of points in counter-clockwise order
+typedef std::vector<uint32_t> PolygonSimple;
+
+// A complex polygon is a list of simple polygons
+// holes are represented by polygons in clockwise order
+typedef std::vector<PolygonSimple> PolygonComplex;
+
+typedef std::vector<uint32_t> Triangles;
+
+enum VertexType {
+    Start,
+    Split,
+    End,
+    Merge,
+    Regular,
+};
+
+struct Vertex {
+    uint32_t index;
+    VertexType type;
+};
+
+struct EdgeEntry {
+    Vertex origin;
+    uint32_t prev;
+    uint32_t next;
+};
+
+double absoluteAngle(Point2D p1, Point2D p2) { return atan2(p2.x - p1.x, p2.y - p1.y); }
+
+VertexType toVertexType(Point2D prev, Point2D current, Point2D next) {
```

Comment:
Parameters that are not trivial data types (int, char, double) should be passed as reference or const ref. Especially when the function is called often